accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,969 @@
+type.
+
+*For example, in `inline class V { final R id; ...}`, `id` has type
+`R`, and `this` has type `V`.*
+
+Again, let `V` be an inline type of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
+and let `R` be the corresponding instantiated representation type.
+If `R` is not an inline type then we say that `V` is an inline type
+at level zero. If `R` is an inline type at level _k_ then we say that
+`V` is an inline type at level _k + 1_.
+A compile-time error occurs if the level of `V` is undefined.
+
+*In other words, cycles are not allowed.*
+
+For every inline class declaration _DV_ named `V`, it is a
+compile-time error if the inline type level of the raw type `V`
+is undefined.
+
+*In other words, we check for this kind of cycles on every non-generic
+inline class declaration using the type directly (even in the case
+where nobody uses that type), and we check for this kind of cycles on
+the instantiation-to-bound of every generic inline class declaration
+(again, even when nobody uses it). Finally, we check for this kind of
+cycles with every parameterized type `V<T1..Tk>` that actually
+occurs in the code which is being analyzed/compiled.*
+
+An inline class declaration _DV_ named `Inline` may declare one or
+more constructors. A constructor which is declared in an inline class
+declaration is also known as an _inline class constructor_.
+
+*The purpose of having an inline class constructor is that it bundles
+an approach for building an instance of the representation type of an
+inline declaration _DV_ with _DV_ itself, which makes it easy to
+recognize that this is a way to obtain a value of that inline type. It
+can also be used to verify that an existing object (provided as an
+actual argument to the constructor) satisfies the requirements for
+having that inline type.*
+
+A compile-time error occurs if an inline class constructor includes a
+superinitializer. *That is, a term of the form `super(...)` or
+`super.id(...)` as the last element of the initializer list.*
+
+*In the body of a generative inline class constructor, the static type
+of `this` is the same as it is in any instance member of the inline,
+that is, `Inline<X1 .. Xk>`, where `X1 .. Xk` are the type parameters
+declared by `Inline`.*
+
+An instance creation expression of the form
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;(...)</code>
+or
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;.name(...)</code>
+is used to invoke these constructors, and the type of such an expression is
+<code>Inline&lt;T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>.
+
+*In short, inline class constructors appear to be very similar to
+constructors in regular classes, and they correspond to the situation
+where the enclosing class has a single, non-late, final instance
+variable, which is initialized according to the normal rules for
+constructors (in particular, it must occur by means of `this.id` or in
+an initializer list).*
+
+
```

Comment:
Is there a missing error around declaring the Object methods?

Reply:
The previous proposals actually allowed shadowing `Object` methods. I've added a rule to make it an error, as agreed yesterday.